File: week-1/week1-1.py
import sys
n = eval(input())
string = sys.stdin.readline()
strdict = {}
ans = 1
last = 0
maxlen = 1
for i, item in enumerate(string):
    if item not in strdict or strdict[item] < last:
        if i == 0:
            maxlen = 1
        else:
            maxlen +=  1
        strdict[item] = i
        ans = max(ans, maxlen)
    else:
        last = strdict[item] + 1
        strdict[item] = i
        maxlen = i + 1 - last

print(ans)

File: week-1/week1-1.cpp
#include<iostream>
#include<map>
using namespace std;
int last_ascii[200];
int vis_ascii[200];
int main()
{
    int n;
    char a;
    cin >> n;
    int cnt = 0;
    int last = 0, maxlen = 0, ans = 1;
    getchar();
    while(cnt < n)
    {
        a = getchar();
        if(!vis_ascii[int(a)] || last_ascii[int(a)] < last)
        {
            vis_ascii[int(a)] = 1;
            maxlen++;
            last_ascii[int(a)] = cnt;
            ans = max(ans, maxlen);
        }
        else
        {
            last = last_ascii[int(a)] + 1;
            last_ascii[int(a)] = cnt;
            maxlen = cnt + 1 - last;
        }
        cnt++;
    }
    cout<<ans<<endl;
    return 0;
}